File: CLDemo-OC/CLDemo-OC/Class/Base/CLController/CLController.m
//
//  CLController.m
//  CLDemo
//
//  Created by JmoVxia on 2016/11/17.
//  Copyright © 2016年 JmoVxia. All rights reserved.
//

#import "CLController.h"
#import "CLNavigationController.h"

@interface CLController ()<UIGestureRecognizerDelegate>

@property (nonatomic, strong) UILabel *titleLabel;

@end

@implementation CLController

- (UILabel *)titleLabel {
    if (!_titleLabel) {
        _titleLabel = [[UILabel alloc] init];
        _titleLabel.font = [UIFont boldSystemFontOfSize:18];
        _titleLabel.textColor = [UIColor orangeColor];
    }
    return _titleLabel;
}

// MARK: - view生命周期

-(void)viewDidLoad {
    [super viewDidLoad];
    [self initSubViews];
    [self makeConstraints];
}

-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [(CLNavigationController *)self.navigationController setNavigationBarBackgroundColor:self.navigationBarBackgroundColor];
    [self.navigationController setNavigationBarHidden:self.isHiddenNavigationBar animated:animated];
}

-(void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
}

-(void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
}

-(void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
}

-(void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
}
-(void)initSubViews {
    self.hiddenStatusBar = NO;
    self.hiddenNavigationBar = NO;
    self.sideSlipBack = YES;
    self.statusBarStyle = UIStatusBarStyleDefault;
    self.navigationBarBackgroundColor = [UIColor whiteColor];
    self.extendedLayoutIncludesOpaqueBars = NO;
    self.view.backgroundColor = [UIColor whiteColor];
    self.navigationItem.titleView = self.titleLabel;
    id target = self.navigationController.interactivePopGestureRecognizer.delegate;
    UIPanGestureRecognizer *pan = [[UIPanGestureRecognizer alloc] initWithTarget:target action:NSSelectorFromString(@"handleNavigationTransition:")];
    pan.delegate = self;
    self.navigationController.interactivePopGestureRecognizer.enabled = NO;
    [self.view addGestureRecognizer:pan];
}

-(void)makeConstraints {
}

// MARK: - 重写父类方法

-(BOOL)shouldAutorotate {
    return YES;
}

-(UIInterfaceOrientationMask)supportedInterfaceOrientations {
    return UIInterfaceOrientationMaskPortrait;
}

-(UIInterfaceOrientation)preferredInterfaceOrientationForPresentation {
    return UIInterfaceOrientationPortrait;
}

-(UIStatusBarStyle)preferredStatusBarStyle {
    return self.statusBarStyle;
}

-(BOOL)prefersStatusBarHidden {
    return self.isHiddenStatusBar;
}

-(UIUserInterfaceStyle)overrideUserInterfaceStyle API_AVAILABLE(ios(13.0)) {
    return UIUserInterfaceStyleLight;
}

// MARK: - 手势代理

-(BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldReceiveTouch:(UITouch *)touch {
    if (!self.isSideSlipBack) {
        return NO;
    }
    BOOL isResponds = [self.navigationController.interactivePopGestureRecognizer.delegate respondsToSelector:NSSelectorFromString(@"handleNavigationTransition:")];
    if (!isResponds) {
        return NO;
    }
    return self.navigationController.childViewControllers.count > 1;
}

// MARK: - public方法

-(void)back {
    if (self.navigationController.presentingViewController != nil && self.navigationController.viewControllers.count == 1) {
        [self dismissViewControllerAnimated:YES completion:nil];
    } else {
        [self.navigationController popViewControllerAnimated:YES];
    }
}

-(void)updateTitleLabel:(void (^)(UILabel *))viewCallback {
    if (viewCallback) {
        viewCallback(self.titleLabel);
        [self.titleLabel sizeToFit];
        self.navigationItem.titleView = self.titleLabel;
    }
}

// MARK: - 重写set,get方法

-(void)setTitleText:(NSString *)titleText {
    if (titleText.length > 0) {
        self.titleLabel.text = titleText;
    }
}

- (void)setStatusBarStyle:(UIStatusBarStyle)statusBarStyle {
    _statusBarStyle = statusBarStyle;
    [self setNeedsStatusBarAppearanceUpdate];
}

- (void)setHiddenStatusBar:(BOOL)hiddenStatusBar {
    _hiddenStatusBar = hiddenStatusBar;
    [self setNeedsStatusBarAppearanceUpdate];
}

- (void)setHiddenNavigationBar:(BOOL)hiddenNavigationBar {
    _hiddenNavigationBar = hiddenNavigationBar;
    [self.navigationController setNavigationBarHidden:_hiddenNavigationBar animated:YES];
}

- (void)setNavigationBarBackgroundColor:(UIColor *)navigationBarBackgroundColor {
    _navigationBarBackgroundColor = navigationBarBackgroundColor;
    [(CLNavigationController *)self.navigationController setNavigationBarBackgroundColor:_navigationBarBackgroundColor];
}

-(CGFloat)navigationBarHeight {
    return self.navigationController.navigationBar.bounds.size.height;
}

-(CGFloat)tabBarHeight {
    return self.tabBarController.tabBar.frame.size.height;
}

-(NSString *)titleText {
    return self.titleLabel.text;
}

@end

File: CLDemo-Swift/CLDemo-Swift/Class/Base/CLController/CLController.swift
//
//  CLController.swift
//  CLDemo-Swift
//
//  Created by Chen JmoVxia on 2021/6/22.
//

import UIKit


//MARK: - JmoVxia---类-属性
class CLController: UIViewController {
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    deinit {
        CLLog("\(self.classForCoder) deinit")
    }
    ///自定义返回响应
    var customBackActionCallback: (() -> (Bool))? {
        didSet {
            (navigationController as? CLNavigationController)?.customBackActionCallback = customBackActionCallback
        }
    }
    ///是否允许侧滑
    var isCanSideBack: Bool = true
    ///状态栏颜色
    var statusBarStyle: UIStatusBarStyle = .default {
        didSet {
            if oldValue != statusBarStyle {
                setNeedsStatusBarAppearanceUpdate()
            }
        }
    }
    ///是否隐藏状态栏
    var isHiddenStatusBar: Bool = false {
        didSet {
            setNeedsStatusBarAppearanceUpdate()
        }
    }
    ///导航条高度
    var navigationBarHeight: CGFloat {
        return navigationController?.navigationBar.bounds.height ?? 0
    }
    ///顶部文字
    var titleText: String? {
        return titleLabel.text
    }
    ///转子标识符
    private var identifier: String?
    private lazy var titleLabel: UILabel = {
       let view = UILabel()
        view.font = PingFangSCBold(18)
        view.textColor = .hex("333333")
        return view
    }()
}
//MARK: - JmoVxia---生命周期
extension CLController {
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
    }
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
    }
    override func viewDidLoad() {
        super.viewDidLoad()
        initUI()
        makeConstraints()
    }
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
    }
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
    }
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
    }
}
//MARK: - JmoVxia---布局
private extension CLController {
    func initUI() {
        view.backgroundColor = .white
        navigationItem.titleView = titleLabel
        let target = navigationController?.interactivePopGestureRecognizer?.delegate
        let pan = UIPanGestureRecognizer.init(target: target, action: Selector(("handleNavigationTransition:")))
        pan.delegate = self
        navigationController?.interactivePopGestureRecognizer?.isEnabled = false
        view.addGestureRecognizer(pan)
    }
    func makeConstraints() {
    }
}
//MARK: - JmoVxia---override
extension CLController {
    override var shouldAutorotate: Bool {
        return false
    }
    override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        return .portrait
    }
    override var preferredInterfaceOrientationForPresentation: UIInterfaceOrientation {
        return .portrait
    }
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return statusBarStyle
    }
    override var prefersStatusBarHidden: Bool {
        return isHiddenStatusBar
    }
    @available(iOS 13.0, *) override var overrideUserInterfaceStyle: UIUserInterfaceStyle {
        set {
            
        }
        get {
            return .light
        }
    }
}
//MARK: - JmoVxia---公共方法
extension CLController {
    ///更新顶部label
    func updateTitleLabel(_ viewCallback:@escaping ((UILabel) -> ())) {
        DispatchQueue.main.async {
            viewCallback(self.titleLabel)
            self.titleLabel.sizeToFit()
        }
    }
    ///显示加载动画
    func showProgress() {
        identifier = CLPopupManager.showLoading()
    }
    ///隐藏加载动画
    func hiddenProgress() {
        guard let identifier = identifier else { return }
        CLPopupManager.dismiss(identifier)
    }
    ///返回
    func back() {
        (navigationController as? CLNavigationController)?.backItemAction()
    }
}
//MARK: - JmoVxia---UIGestureRecognizerDelegate
extension CLController :UIGestureRecognizerDelegate {
    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldReceive touch: UITouch) -> Bool {
        if isCanSideBack {
            let isResponds: Bool = navigationController?.interactivePopGestureRecognizer?.delegate?.responds(to: Selector(("handleNavigationTransition:"))) ?? false
            if  isResponds {
                let count = navigationController?.children.count ?? 0
                return count > 1
            }
        }else {
            return false
        }
        return false
    }
}
